docs/control-flow.md

```diff
@@ -0,0 +1,363 @@
+  text_representation:
+    extension: .md
+    format_name: myst
+    format_version: 0.13
+    jupytext_version: 1.16.4
+kernelspec:
+  display_name: Python 3
+  language: python
+  name: python3
+---
+
++++ {"id": "rg4CpMZ8c3ri"}
+
+(control-flow)=
+# Control flow and logical operators with JIT
+
+<!--* freshness: { reviewed: '2024-11-11' } *-->
+
+When executing eagerly (outside of `jit`), JAX code works with Python control flow and logical operators just like Numpy code. Using control flow and logical operators with `jit` is more complicated. 
+
+In a nutshell, Python control flow and logical operators are evaluated at JIT compile time, such that the compiled function represents a single path through the [control flow graph](https://en.wikipedia.org/wiki/Control-flow_graph) (logical operators affect the path via short-circuiting). If the path depends on the values of the inputs, the function (by default) cannot be JIT compiled. The path may depend on the shape or dtype of the inputs, and the function is re-compiled every time it is called on an input with a new shape or dtype.
+
+```{code-cell}
+from jax import grad, jit
+import jax.numpy as jnp
+```
+
+For example, this works:
+
+```{code-cell}
+:id: OZ_BJX0CplNC
+:outputId: 60c902a2-eba1-49d7-c8c8-2f68616d660c
+
+@jit
+def f(x):
+  for i in range(3):
+    x = 2 * x
+  return x
+
+print(f(3))
+```
+
++++ {"id": "22RzeJ4QqAuX"}
+
+So does this:
+
+```{code-cell}
+:id: pinVnmRWp6w6
+:outputId: 25e06cf2-474f-4782-af7c-4f5514b64422
+
+@jit
+def g(x):
+  y = 0.
+  for i in range(x.shape[0]):
+    y = y + x[i]
+  return y
+
+print(g(jnp.array([1., 2., 3.])))
+```
+
++++ {"id": "TStltU2dqf8A"}
+
+But this doesn't, at least by default:
+
+```{code-cell}
+:id: 9z38AIKclRNM
+:outputId: 38dd2075-92fc-4b81-fee0-b9dff8da1fac
+:tags: [raises-exception]
+
+@jit
+def f(x):
+  if x < 3:
+    return 3. * x ** 2
+  else:
+    return -4 * x
+
+# This will fail!
+f(2)
+```
+
+Neither does this:
+
+```{code-cell}
+:tags: [raises-exception]
+
+@jit
+def g(x):
+  return (x > 0) and (x < 3)
+
+# This will fail!
+g(2)
+```
+
++++ {"id": "pIbr4TVPqtDN"}
+
+__What gives!?__
+
+When we `jit`-compile a function, we usually want to compile a version of the function that works for many different argument values, so that we can cache and reuse the compiled code. That way we don't have to re-compile on each function evaluation.
+
+For example, if we evaluate an `@jit` function on the array `jnp.array([1., 2., 3.], jnp.float32)`, we might want to compile code that we can reuse to evaluate the function on `jnp.array([4., 5., 6.], jnp.float32)` to save on compile time.
+
+To get a view of your Python code that is valid for many different argument values, JAX traces it on _abstract values_ that represent sets of possible inputs. There are [multiple different levels of abstraction](https://github.com/jax-ml/jax/blob/main/jax/_src/abstract_arrays.py), and different transformations use different abstraction levels.
```

Comment:
As of last week, there are no longer multiple levels of abstraction (it's basically just `ShapedArray` now) so we could probably remove that line. We also need to adjust the next sentence.